### PR TITLE
Type hash distribution during discovery (rep2011)

### PR DESCRIPTION
**Pull request by emersonknapp**
Part of https://github.com/ros2/ros2/issues/1159

Depends on https://github.com/ros2/rosidl/pull/729
Depends on https://github.com/ros2/rmw_dds_common/pull/70

Features:
* Retrieve type version hash from generated code, and use the USER_DATA QoS policy to distribute it for all readers/writers. Format of the USER_DATA uses common utility functions in `rmw_dds_common`
* On discovering new endpoints, parse the USER_DATA to set `rmw_topic_endpoint_info_t.type_hash`

### CODE REVIEW

File: rmw_connextdds_common/src/common/rmw_graph.cpp [hunk 1]
```diff
@@ -982,12 +988,18 @@ rmw_connextdds_graph_add_remote_entity(
     return RMW_RET_OK;
   }
 
+  const uint8_t * user_data_data = DDS_OctetSeq_get_contiguous_buffer(&user_data->value);
+  const size_t user_data_size = DDS_OctetSeq_get_length(&user_data->value);
+  rosidl_type_hash_t type_hash = rmw_dds_common::parse_type_hash_from_user_data(
+    user_data_data, user_data_size);
```
Comment (clalancette) [inline rmw_connextdds_common/src/common/rmw_graph.cpp:994]: This needs to be updated with the new API, and do error checking.

File: rmw_connextdds_common/src/common/rmw_impl.cpp [hunk 2]
```diff
@@ -454,6 +454,13 @@ rmw_connextdds_get_readerwriter_qos(
   }
 #endif /* RMW_CONNEXT_DDS_API == RMW_CONNEXT_DDS_API_PRO */
 
+  auto user_data_str = rmw_dds_common::encode_type_hash_for_user_data_qos(
+    *type_support->type_hash());
```
Comment (clalancette) [inline rmw_connextdds_common/src/common/rmw_impl.cpp:458]: This needs to be updated with the new API and do error checking.  Also, let's not use `auto` here.

**Review by clalancette — CHANGES_REQUESTED** [review]

File: rmw_connextdds_common/src/common/rmw_graph.cpp [hunk 3]
```diff
@@ -727,6 +728,7 @@ rmw_connextdds_graph_add_entityEA(
   const DDS_GUID_t * const dp_guid,
   const char * const topic_name,
   const char * const type_name,
+  const rosidl_type_hash_t * type_hash,
```
[the remark this answers is not in the document]
Comment (emersonknapp) [reply rmw_connextdds_common/src/common/rmw_graph.cpp]: Changed to a const ref instead of ptr to obviate the need

**Review by clalancette — CHANGES_REQUESTED** [review]
Two small changes, this otherwise looks good to me.

**Comment by emersonknapp** [comment]
Gist: https://gist.githubusercontent.com/emersonknapp/597121a37b82d9418b14472de760c8f0/raw/b0412d81c2ae2368cb592c810496d6f92650ad68/ros2.repos
BUILD args: 
TEST args: 
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/11623
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/18247/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/12780/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/18921/)

**Review by clalancette — APPROVED** [review]

**Comment by emersonknapp** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/18261/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/12797/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/18938/)

**Review by wjwwood — APPROVED** [review]

**Review by clalancette — APPROVED** [review]

**Comment by clalancette** [comment]
CI is in https://github.com/ros2/rmw_dds_common/pull/70#issuecomment-1483326412

**Pull request merged**